dipy/denoise/fast_lpca.pyx

```diff
@@ -0,0 +1,270 @@
+import numpy as np
+cimport cython
+cimport numpy as cnp
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef void fast_vecmat_mul(double[:] x, double[:, :] A, double[:] out) nogil:
+    '''
+    Fast vector multiplication with matrix
+    out = xA
+    '''
+    cdef:
+        cnp.npy_intp i, j
+        double s
+        cnp.npy_intp n0 = x.shape[0]
+        cnp.npy_intp An1 = A.shape[1]
+
+    for i in range(An1):
+        s = 0
+        for j in range(n0):
+            s += x[j] * A[j, i]
+
+        out[i] = s
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef void update_outer_prod(double[:,:] out, double[:] x, int utype)nogil:
+    cdef:
+        cnp.npy_intp i, j, n
+    n = x.shape[0]
+    if utype==0:
+        for i in range(n):
+            for j in range(n):
+                out[i,j] = x[i] * x[j]
+    elif utype==1:
+        for i in range(n):
+            for j in range(n):
+                out[i,j] += x[i] * x[j]
+    elif utype==-1:
+        for i in range(n):
+            for j in range(n):
+                out[i,j] -= x[i] * x[j]
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef void update_vector(double[:] out, double[:] x, int utype)nogil:
+    cdef:
+        cnp.npy_intp i, n
+    n = x.shape[0]
+    if utype==0:
+        for i in range(n):
+            out[i] = x[i]
+    elif utype==1:
+        for i in range(n):
+            out[i] += x[i]
+    elif utype==-1:
+        for i in range(n):
+            out[i] -= x[i]
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef void update_matrix(double[:,:] out, double[:,:] cov, int utype)nogil:
+    cdef:
+        cnp.npy_intp i, j, n
+    n = cov.shape[0]
+    if utype==0:
+        for i in range(n):
+            for j in range(n):
+                out[i,j] = cov[i, j]
+    elif utype==1:
+        for i in range(n):
+            for j in range(n):
+                out[i,j] += cov[i, j]
+    elif utype==-1:
+        for i in range(n):
+            for j in range(n):
+                out[i,j] -= cov[i, j]
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+def fast_lpca(double[:, :, :, :] I, int radius, double[:, :, :, :] sigma):
```

Comment:
no we are not, the function is having 4D for convenience, but the sigma at all directions are same